### DISCUSSION TITLE
How to handle multiple commits within a single savepoint in 2.0-style?

### DISCUSSION THREAD

**Discussion by onecrayon** [sqlalchemy/sqlalchemy#6901] [Usage Questions] [answered]
One of my databases creates universal UUIDs for some of its objects that are then used as external IDs for those objects; e.g.:

```python
class GlobalUuid(DeclarativeBase):
    __tablename__ = "global_uuids"
    uuid = Column(
        UUID(as_uuid=True), primary_key=True, server_default=func.gen_random_uuid()
    )

@declarative_mixin
class KeyedBase:
    @declared_attr
    def uuid(cls):
        return Column(
            UUID(as_uuid=True),
            ForeignKey(GlobalUuid.uuid, ondelete="cascade"),
            unique=True,
        )
```

However, I'm running into trouble when I try to create these objects. I was planning to create a nested savepoint off the current session, insert/commit the GlobalUuid (if necessary), then insert/commit the new object. E.g.:

```python
class KeyedBase:
    # ...stuff above here...
    async def commit_to_session(cls, session: AsyncSession, keyed_object):
       try:
            async with session.begin_nested():
                # Grab a new global UUID, if necessary
                if not keyed_object.uuid:
                    global_uuid = GlobalUuid(data_type=cls.__tablename__)
                    session.add(global_uuid)
                    await session.commit()
                    keyed_object.uuid = global_uuid.uuid
                # Then add our object to the session and commit
                session.add(keyed_object)
                await session.commit()
                return keyed_object
        except:
            logger.error(
                f"Global UUID creation failed for {cls.__tablename__}!", exc_info=True
            )
            raise
```

However, as soon as I commit the GlobalUuid, the savepoint is closed and the second commit throws an error because it's within a dead transaction. I discovered this is expected behavior per [the docs](https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#commit-as-you-go):

> Both Session and Connection feature Connection.commit() and Connection.rollback() methods. Using SQLAlchemy 2.0-style operation, these methods affect the **outermost** transaction in all cases.

This presents me with a conundrum, since I don't want the GlobalUuid to stick around if some exception is thrown when creating whatever inherits from KeyedBase. Do I just need to use try/except and explicitly delete the UUID object if it was created (basically a by-hand rollback)? Or is there a way in 2.0-style to use multiple commits within a single savepoint?

**Comment by zzzeek** [comment, marked as answer]
im not sure of the whole issue here but what's immediately wrong is the use of "with session.begin()" with a "session.commit()" inside the block.  this block commits automatically when it resolves.  if you need to commit as you go, then don't use the "with" block, but looking at the code I would say just remove those .commit() calls, im not sure what they're trying to accomplish.

**Reply by onecrayon** [reply]
@zzzeek I need to commit the GlobalUuid object to the database (since that's where the UUID is generated). So you're recommending something like this instead?

```python
try:
    nested_session = session.begin_nested()
    # perform all adding/committing on nested_session here
except:
    nested_session.rollback()
```

The short answer to what I'm trying to do is that flushing this stuff to the database could be occurring within a broader session (I don't know for sure where the utility function will be used), which is why I wanted to segregate it all in a savepoint.

**Reply by zzzeek** [reply]
so the control flow you need here is if the global UUID INSERT fails due to a unique constraint, you need the program to continue.  so by definition you need a try/except.   whether you use the "with" is still a choice.  general form is (add async keywords as needed):

```py
with session.begin():
    try:
        with session.begin_nested():
            session.add(GlobalUUIDThatMightExist())
    except IntegrityError:
        log.debug("global uuid already existed")
        # the savepoint would be rolled back here automatically

     # do other things in transaction
```

beyond that, im still not sure this fits what you need since you're making this new "keyed_object" which still needs to have a GlobalUuid object present.   Which means i don't see how you avoid performing a SELECT for this GlobalUuid object in any case, if the INSERT fails, you need to have that row.

**Reply by onecrayon** [reply]
I'm using PostgreSQL and the GlobalUuid table has `__mapper_args__ = {"eager_defaults": True}`...which I apparently managed to copy and paste out of existence in the example code above. 🤦 

So basically, I want the whole transaction to fail and get rolled back if either the GlobalUuid fails, or if the insertion of the new keyed object fails. This is why I was trying to wrap both commits in the same savepoint (I believe I had a similar pattern working in 1.0-style in the past, which is why I started there).

Incidentally, thank you for your ongoing help. I wasn't expecting a reply from the maintainer. 😄

**Reply by CaselIT** [reply]
To generate the uuid server side you can also just flush after adding global uuid, then you can add the other objects using the generated id.

**Reply by onecrayon** [reply]
Arg, of course flushing is the answer! I don't know why I completely spaced on that being a thing. Thank you so much! Here's an example of the final logic that I landed on (though I still need to identify the likely exceptions and catch those specifically):

```python
try:
    async with session.begin_nested():
        # Grab a new global UUID, if necessary
        if not keyed_object.uuid:
            global_uuid = GlobalUuid()
            session.add(global_uuid)
            await session.flush()
            keyed_object.uuid = global_uuid.uuid
        # Then add our object to the session (it is committed automatically when we exit the context manager)
        session.add(keyed_object)
    return keyed_object
except:
    # error reporting here
```

Thank you very much for both of your help!
